include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -0,0 +1,134 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_PSTL_OFFLOAD_INTERNAL_USM_MEMORY_REPLACEMENT_COMMON_H
+#define _ONEDPL_PSTL_OFFLOAD_INTERNAL_USM_MEMORY_REPLACEMENT_COMMON_H
+
+#include <sycl/sycl.hpp>
+#include <cstddef>
+#include <cstdlib>
+#include <cassert>
+#include <cstring>
+#include <limits>
+
+#if __linux__
+#include <dlfcn.h>
+#include <unistd.h>
+#endif // __linux__
+
+namespace __pstl_offload {
+
+inline constexpr std::size_t __uniq_type_const = 0x23499abc405a9bccLLU;
+inline constexpr std::size_t __header_offset = 32;
+
+struct alignas(__header_offset) __block_header {
+    std::size_t _M_uniq_const;
+    void* _M_original_pointer;
+    sycl::device* _M_device;
+    std::size_t _M_requested_number_of_bytes;
+}; // struct __block_header
+
+static_assert(sizeof(__block_header) <= __header_offset);
```

Comment:
I guess it can be == now, but does it make sense in that case?

Reply:
Interesting. According to my understanding, `alignas` is guarantee something about object's address in cast of the object allocation (btw that is not our case, as we cast, not allocate) and nothing about the object's sizeof. Is it incorrect? Are you sure that sizeof guarantees are provided?

Reply:
`alignas` would affect `sizeof` as well. Consider allocating an array of objects - the block of memory would be continuous and since the alignment requirement for each element in the array should be satisfied, the objects sizeof may be increased.
Like in [this example](https://godbolt.org/z/8xeov8s3T).
In our case, we decided to add `alignas` to guarantee that `sizeof(__block_header)` is at least `__header_offset` or a multiple of it.

Regarding the static assertion, I guess it can be removed.

Reply:
Thanks, but I don't mean an implementation, I was interested in some guarantee from the standard.

Reply:
Changed as discussed internally with Ruslan